[TACHYON-1318] Add a findbug rule of finding unused variables

RongGu opened the pull request:
Add findbug rule for unread/unused private variables. The public/protected ones are allowed since they may be used with classes not seen as part of the analysis.

Also, this PR suppresses the found bugs with `@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("URF_UNREAD_FIELD")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/7274/
Test PASSed.

aaudiber commented on servers/src/main/java/tachyon/worker/nio/NIODataServer.java:
```diff
@@ -61,6 +61,7 @@
 
   // Instance of TachyonConf
+  @edu.umd.cs.findbugs.annotations.SuppressFBWarnings("URF_UNREAD_FIELD")
   private final TachyonConf mTachyonConf;
```
We should be able to just remove this field

RongGu replied:
Sure, I'll remove it.

aaudiber commented:
Thanks for the improvement to Findbugs coverage! I've left a few comments. Sorry some of them landed on a previous commit - was that because of a commit `--amend` and `push -f`?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/7275/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/7276/<h2>Build result: FAILURE</span></h2>[...truncated 4677 lines...][JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/target/tachyon-client-unshaded-0.9.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-client-unshaded/0.9.0-SNAPSHOT/tachyon-client-unshaded-0.9.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/pom.xml to org.tachyonproject/tachyon-servers/0.9.0-SNAPSHOT/tachyon-servers-0.9.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.9.0-SNAPSHOT.jar to org.tachyonproject/tachyon-servers/0.9.0-SNAPSHOT/tachyon-servers-0.9.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.9.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-servers/0.9.0-SNAPSHOT/tachyon-servers-0.9.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.9.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-servers/0.9.0-SNAPSHOT/tachyon-servers-0.9.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/target/tachyon-servers-0.9.0-SNAPSHOT-tests.jar to org.tachyonproject/tachyon-servers/0.9.0-SNAPSHOT/tachyon-servers-0.9.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/pom.xml to org.tachyonproject/tachyon-underfs-hdfs/0.9.0-SNAPSHOT/tachyon-underfs-hdfs-0.9.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.9.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-hdfs/0.9.0-SNAPSHOT/tachyon-underfs-hdfs-0.9.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.9.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-hdfs/0.9.0-SNAPSHOT/tachyon-underfs-hdfs-0.9.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.9.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-hdfs/0.9.0-SNAPSHOT/tachyon-underfs-hdfs-0.9.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/pom.xml to org.tachyonproject/tachyon-parent/0.9.0-SNAPSHOT/tachyon-parent-0.9.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/dependency-reduced-pom.xml to org.tachyonproject/tachyon-client/0.9.0-SNAPSHOT/tachyon-client-0.9.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.9.0-SNAPSHOT.jar to org.tachyonproject/tachyon-client/0.9.0-SNAPSHOT/tachyon-client-0.9.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.9.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-client/0.9.0-SNAPSHOT/tachyon-client-0.9.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/target/tachyon-client-0.9.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-client/0.9.0-SNAPSHOT/tachyon-client-0.9.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/pom.xml to org.tachyonproject/tachyon-shell/0.9.0-SNAPSHOT/tachyon-shell-0.9.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/target/tachyon-shell-0.9.0-SNAPSHOT.jar to org.tachyonproject/tachyon-shell/0.9.0-SNAPSHOT/tachyon-shell-0.9.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/target/tachyon-shell-0.9.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-shell/0.9.0-SNAPSHOT/tachyon-shell-0.9.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/target/tachyon-shell-0.9.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-shell/0.9.0-SNAPSHOT/tachyon-shell-0.9.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.tachyonproject/tachyon-underfs-glusterfs/0.9.0-SNAPSHOT/tachyon-underfs-glusterfs-0.9.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.9.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.9.0-SNAPSHOT/tachyon-underfs-glusterfs-0.9.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.9.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.9.0-SNAPSHOT/tachyon-underfs-glusterfs-0.9.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.9.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.9.0-SNAPSHOT/tachyon-underfs-glusterfs-0.9.0-SNAPSHOT-javadoc.jarchannel stopped
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/7278/
Test PASSed.

RongGu commented:
@aaudiber , thanks and I've addressed the comments.
Remove unused fields. Use `@SuppressFBWarnings` only. Use the 'compile' scope which is more understandable and add description for the dependency.

jsimsa commented on pom.xml:
```diff
@@ -283,6 +283,18 @@
         <version>${powermock.version}</version>
         <scope>test</scope>
       </dependency>
+
```
do we need this in the top-level package if the annotation is only used in the servers module?

swatDong replied:
Currently only servers use. But this package can be used to manually handle any special findbug cases.
Both is OK for me.

RongGu replied:
@aaudiber , any suggestions about putting to the top-level package?

aaudiber replied:
I think top-level is fine as long as the scope keeps the dependency out of the jar.

RongGu commented:
@aaudiber @jsimsa , thanks. I've changed the scope to `provided` and kept it still in top-level.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/7293/
Test PASSed.

jsimsa commented:
LGTM

aaudiber commented:
LGTM
